extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -303,70 +252,138 @@ export class AmpA4A extends AMP.BaseElement {
     this.adPromise_ = viewerFor(this.win).whenFirstVisible()
-      // This block returns the ad URL, if one is available.
-      /** @return {!Promise<?string>} */
-      .then(() => {
-        checkStillCurrent(promiseId);
-        return this.getAdUrl();
-      })
-      // This block returns the (possibly empty) response to the XHR request.
-      /** @return {!Promise<?Response>} */
-      .then(adUrl => {
-        checkStillCurrent(promiseId);
-        this.adUrl_ = adUrl;
-        return adUrl && this.sendXhrRequest_(adUrl);
-      })
-      // The following block returns either the response (as a {bytes, headers}
-      // object), or null if no response is available / response is empty.
-      /** @return {!Promise<?{bytes: !ArrayBuffer, headers: !Headers}>} */
-      .then(fetchResponse => {
-        checkStillCurrent(promiseId);
-        if (!fetchResponse || !fetchResponse.arrayBuffer) {
-          return null;
-        }
-        // Note: Resolving a .then inside a .then because we need to capture
-        // two fields of fetchResponse, one of which is, itself, a promise,
-        // and one of which isn't.  If we just return
-        // fetchResponse.arrayBuffer(), the next step in the chain will
-        // resolve it to a concrete value, but we'll lose track of
-        // fetchResponse.headers.
-        return fetchResponse.arrayBuffer().then(bytes => {
-          return {
-            bytes,
-            headers: fetchResponse.headers,
-          };
-        });
-      })
-      // This block returns the ad creative and signature, if available; null
-      // otherwise.
-      /**
-       * @return {!Promise<?{creative: !ArrayBuffer, signature: !ArrayBuffer}>}
-       */
-      .then(responseParts => {
-        checkStillCurrent(promiseId);
-        return responseParts && this.extractCreativeAndSignature(
-                responseParts.bytes, responseParts.headers);
-      })
-      // This block returns the ad creative if it exists and validates as AMP;
-      // null otherwise.
-      /** @return {!Promise<?string>} */
-      .then(creativeParts => {
-        checkStillCurrent(promiseId);
-        return creativeParts && this.validateAdResponse_(
-            creativeParts.creative, creativeParts.signature);
-      })
-      // This block returns true iff the creative was rendered in the shadow
-      // DOM.
-      /** @return {!Promise<!boolean>} */
-      .then(creative => {
-        checkStillCurrent(promiseId);
-        // Note: It's critical that #maybeRenderAmpAd_ be called
-        // on precisely the same creative that was validated
-        // via #validateAdResponse_.  See GitHub issue
-        // https://github.com/ampproject/amphtml/issues/4187
-        return creative && this.maybeRenderAmpAd_(creative);
-      })
-      .catch(error => this.promiseErrorHandler_(error));
+    // This block returns the ad URL, if one is available.
+    /** @return {!Promise<?string>} */
+    .then(() => {
```

Comment:
I meant: Continued statements get an additional +4 spaces on top of the indentation of the original line. The statements inside a function param (like the next line), are indented using 2 spaces:

```js
    return originalStatement
        .then(() => {
          return insideStatement;
        })
        .then(() => { // This could be on the previous line.
          return insideStatement;
        });